README.md
# Watson-Framework

    It's elementary my dear Watson

Watson is an easy to use framework designed to get out of your way and let you code your application rather than wrangle with the framework.

For full documentation please see [Read The Docs](https://readthedocs.org/projects/watson-framework/).

#### Build Status

Branch | Status | Coverage
------------ | ------------- | -------------
Master | [![Build Status](https://api.travis-ci.org/bespohk/watson-framework.png?branch=master)](https://travis-ci.org/bespohk/watson-framework) | [![Coverage Status](https://coveralls.io/repos/bespohk/watson-framework/badge.png)](https://coveralls.io/r/bespohk/watson-framework)
Develop | [![Build Status](https://api.travis-ci.org/bespohk/watson-framework.png?branch=develop)](https://travis-ci.org/bespohk/watson-framework) |

[![Pypi](https://pypip.in/v/watson-framework/badge.png)](https://crate.io/packages/watson-framework/)

## Dependencies

* watson-cache
* watson-common
* watson-console
* watson-di
* watson-dev
* watson-events
* watson-filter
* watson-form
* watson-html
* watson-http
* watson-validators

## Issues

.


CHANGES
3.5.0

- BC: No more support for Python 3.3, requiring Enum in watson-forms

3.4.3

- Fixes to component loading for dependencies

3.4.0

- Applications can now be split up into components to better provide organisation (use --component_based)
- New Project no longer splits up configuration into separate files based on server type, prefer ENV vars
- Default .gitignore file generated with New Project

3.3.0

- Views can now also be loaded via the PackageLoader [(package, folder)] format
- Error pages (e.g. 404 and 500) will now respect the ACCEPT header and return the relevant Content-Type
- Default Jinja2 error pages refactored out of .py into separate html files

3.2.1

- Styling and formatting tweaks for debug toolbar

3.2.0

- Upgraded and rethemed debug toolbar, can now be resized
- i18n support via watson.framework.i18n

3.0.0

- BC: Moved commands into own namespace (following watson.console 2.0.0), now accessed via ./console.py dev runserver, and ./console.py project new|routes|test (see ./console.py -h for more information)
- framework.applications.Console now accepts the argv args in the run() method instead of __init__
- Close the debug toolbar entirely if it's closed already
- RouteMatch information included in debug toolbar
- Bugfix for flashmessages in the view
- Jinja2 global functions and classes no longer use the container as a service locator


watson/framework/config.py
# -*- coding: utf-8 -*-
# Default configuration for a Watson application.
# The container itself can be referenced by a simple lambda function such as:
# lambda container: container
#
# Consult the documentation for more indepth setting information.
import os
from watson.framework import events

# Debug settings
debug = {
    'enabled': False,
    'icons_only': False,
    'panels': {
        'watson.framework.debug.panels.Request': {
            'enabled': True
        },
        'watson.framework.debug.panels.Application': {
            'enabled': True
        },
        'watson.framework.debug.panels.Profile': {
            'enabled': True,
            'max_results': 20,
            'sort': 'time',
        },
        'watson.framework.debug.panels.Framework': {
            'enabled': True
        },
        'watson.framework.debug.panels.Logging': {
            'enabled': True
        }
    }
}

# IocContainer settings
dependencies = {
    'definitions': {
        'shared_event_dispatcher':
        {'item': 'watson.events.dispatcher.EventDispatcher'},
        'router': {
            'item': 'watson.routing.routers.DictRouter',
            'init':
            [lambda container: container.get(
             'application.config').get('routes', None)]
        },
        'profiler': {
            'item': 'watson.framework.debug.profilers.Profiler',
            'init':
            [lambda container: container.get(
             'application.config')['debug']['profiling']]
        },
        'exception_handler': {
            'item': 'watson.framework.exceptions.ExceptionHandler',
            'init':
            [lambda container: container.get(
             'application.config').get('debug', {})]
        },
        'jinja2_renderer': {
            'item': 'watson.framework.views.renderers.jinja2.Renderer',
            'init': [
                lambda container: container.get('application.config')[
                    'views']['renderers']['jinja2'].get('config', {}),
                lambda container: container.get('application')
            ]
        },
        'json_renderer': {'item': 'watson.framework.views.renderers.json.Renderer'},
        'xml_renderer': {'item': 'watson.framework.views.renderers.xml.Renderer'},
        'app_dispatch_execute_listener': {
            'item': 'watson.framework.listeners.DispatchExecute',
            'init':
            [lambda container: container.get(
             'application.config')['views']['templates']]
        },
        'app_exception_listener': {
            'item': 'watson.framework.listeners.Exception_',
            'init': [
                lambda container: container.get('exception_handler'),
                lambda container: container.get(
                    'application.config')['views']['templates']
            ]
        },
        'app_render_listener': {
            'item': 'watson.framework.listeners.Render',
            'init':
            [lambda container: container.get('application.config')['views']]
        },
        'translator': {
            'item': 'watson.framework.i18n.translate.Translator',
            'init': [
                lambda container: container.get(
                    'application.config')['i18n']['default_locale'],
                lambda container: container.get(
                    'application.config')['i18n']['package']
            ]
        },
        'mailer_backend': {
            'item': lambda container: container.get(container.get(
                'application.config')['mail']['backend']['class']),
            'init': lambda container: container.get('application.config')['mail']['backend']['options']
        },
        'mailer': {
            'item': 'watson.framework.mail.Mailer',
            'init': [
                lambda container: container.get('mailer_backend'),
                lambda container: container.get(
                    container.get('application.config')['views']['renderers'][container.get(
                        'application.config')['views']['default_renderer']]['name'])
            ]
        }
    }
}

# View settings
views = {
    'default_format': 'html',
    'default_renderer': 'jinja2',
    'renderers': {
        'jinja2': {
            'name': 'jinja2_renderer',
            'config': {
                'extension': 'html',
                'paths': [os.path.join(os.getcwd(), 'views')],
                'packages': [],
                'framework_packages': [
                    ('watson.framework.views.templates', 'html'),
                    ('watson.framework.debug', 'views'),
                ],
                'filters': ['watson.framework.support.jinja2.filters'],
                'globals': ['watson.framework.support.jinja2.globals'],
            }
        },
        'xml': {'name': 'xml_renderer'},
        'json': {'name': 'json_renderer'}
    },
    'templates': {
        '404': 'errors/404',
        '500': 'errors/500'
    }
}

# Logging settings
logging = {
    'callable': 'logging.config.dictConfig',
    'ignore_status': (404,),
    'options': {
        'version': 1,
        'disable_existing_loggers': False,
        'formatters': {
            'verbose': {
                'format': '%(asctime)s - %(name)s - %(levelname)s - %(process)d %(thread)d - %(message)s'
            },
            'simple': {
                'format': '%(asctime)s - %(levelname)s - %(message)s'
            },
        },
        'handlers': {
            'console': {
                'class': 'logging.StreamHandler',
                'level': 'DEBUG',
                'formatter': 'verbose',
                'stream': 'ext://sys.stdout'
            },
        },
        'loggers': {},
        'root': {
            'level': 'DEBUG',
            'handlers': ['console']
        }
    }
}

# Session settings
session = {
    'class': 'watson.http.sessions.File',
    'options': {
        'timeout': 3600
    }
}

# Exceptions
exceptions = {
    'class': 'watson.framework.exceptions.ApplicationError'
}

# Localization
i18n = {
    'default_locale': 'en',
    'package': 'watson.framework.i18n.locales'
}

# Mail
mail = {
    'backend': {
        'class': 'watson.mail.backends.Sendmail',
        'options': {}
    },
}

# Components
components = []

# Application event settings
events = {
    events.EXCEPTION: [('app_exception_listener',)],
    events.INIT: [
        ('watson.framework.logging.listeners.Init', 1),
        ('watson.framework.debug.listeners.Init', 1)
    ],
    events.ROUTE_MATCH: [('watson.framework.listeners.Route',)],
    events.DISPATCH_EXECUTE: [('app_dispatch_execute_listener',)],
    events.RENDER_VIEW: [('app_render_listener',)],
}
